be/src/exprs/agg/window_funnel.h

```diff
@@ -0,0 +1,217 @@
+        if (!sorted && !other_sorted)
+            std::sort(std::begin(events_list), std::end(events_list), ComparePairFirst{});
+        else {
+            const auto begin = std::begin(events_list);
+            const auto middle = std::next(begin, size);
+            const auto end = std::end(events_list);
+
+            if (!sorted) std::stable_sort(begin, middle, ComparePairFirst{});
+
+            if (!other_sorted) std::stable_sort(middle, end, ComparePairFirst{});
+
+            std::inplace_merge(begin, middle, end);
+        }
+
+        sorted = true;
+    }
+
+    void serialize_to_array_column(ArrayColumn* array_column) const {
+        if (!events_list.empty()) {
+            size_t size = events_list.size();
+            DatumArray array;
+            array.reserve(size * 2 + 3);
+            array.emplace_back(window_size);
+            array.emplace_back((int64_t)events_size);
+            array.emplace_back((int64_t)sorted);
+            for (int i = 0; i < size; i++) {
+                array.emplace_back(events_list[i].first);
+                array.emplace_back((int64_t)events_list[i].second);
+            }
+            array_column->append_datum(array);
+        }
+    }
+
+    int32_t get_event_level() const {
+        std::vector<int64_t> events_timestamp(events_size, -1);
+        for (size_t i = 0; i < events_list.size(); i++) {
+            int64_t timestamp = (events_list)[i].first;
+            uint8_t event_idx = (events_list)[i].second;
+
+            // this row match no conditions.
+            if (event_idx == 0) {
+                continue;
+            }
+
+            event_idx -= 1;
+            if (event_idx == 0)
+                events_timestamp[0] = timestamp;
+            else if (events_timestamp[event_idx - 1] >= 0 &&
+                     timestamp <= events_timestamp[event_idx - 1] + window_size) {
+                events_timestamp[event_idx] = events_timestamp[event_idx - 1];
+                if (event_idx + 1 == events_size) return events_size;
+            }
+        }
+        for (size_t event = events_timestamp.size(); event > 0; --event) {
+            if (events_timestamp[event - 1] >= 0) return event;
+        }
+        return 0;
+    }
+};
+
+class WindowFunnelAggregateFunction final
+        : public AggregateFunctionBatchHelper<WindowFunnelState, WindowFunnelAggregateFunction> {
+public:
+    void update(FunctionContext* ctx, const Column** columns, AggDataPtr __restrict state, size_t row_num) const {
+        const auto* window_size_column = down_cast<const Int64Column*>(columns[0]);
+        this->data(state).window_size = window_size_column->get(row_num).get_int64();
+
+        // get timestamp
+        const auto* timestamp_column = down_cast<const TimestampColumn*>(columns[1]);
+        auto tv = timestamp_column->get(row_num).get_timestamp();
+
+        // get event
+        uint8_t event_level = 0;
+        const auto* event_column = down_cast<const ArrayColumn*>(columns[2]);
```

Comment:
with mode parameter array<boolean> is 4rd.
```suggestion
        const auto* event_column = down_cast<const ArrayColumn*>(columns[3]);
```